apiserver/facades/client/application/deployrepository.go

```diff
@@ -80,3 +140,467 @@ func validateDeployFromRepositoryArgs(_ params.DeployFromRepositoryArg) []error
+	}
+	if platform.Channel == "" {
+		mCfg, err := v.model.Config()
+		if err != nil {
+			return corecharm.Platform{}, usedModelDefaultBase, err
+		}
+		if db, ok := mCfg.DefaultBase(); ok {
+			defaultBase, err := coreseries.ParseBaseFromString(db)
+			if err != nil {
+				return corecharm.Platform{}, usedModelDefaultBase, err
+			}
+			platform.OS = defaultBase.OS
+			platform.Channel = defaultBase.Channel.String()
+			usedModelDefaultBase = true
+		}
+	}
+	return platform, usedModelDefaultBase, nil
+}
+
+func (v *deployFromRepositoryValidator) platformFromPlacement(placements []*instance.Placement) (corecharm.Platform, bool, error) {
+	if len(placements) == 0 {
+		return corecharm.Platform{}, false, errors.NotFoundf("placements")
+	}
+	machines := make([]Machine, 0)
+	// Find which machines in placement actually exist.
+	for _, placement := range placements {
+		m, err := v.state.Machine(placement.Directive)
+		if errors.Is(err, errors.NotFound) {
+			continue
+		}
+		if err != nil {
+			return corecharm.Platform{}, false, err
+		}
+		machines = append(machines, m)
+	}
+	if len(machines) == 0 {
+		return corecharm.Platform{}, false, errors.NotFoundf("machines in placements")
+	}
+
+	// Gather platforms for existing machines
+	var platform corecharm.Platform
+	platStrings := set.NewStrings()
+	for _, machine := range machines {
+		b := machine.Base()
+		a, err := machine.HardwareCharacteristics()
+		if err != nil {
+			return corecharm.Platform{}, false, err
+		}
+		platString := fmt.Sprintf("%s/%s/%s", *a.Arch, b.OS, b.Channel)
+		p, err := corecharm.ParsePlatformNormalize(platString)
+		if err != nil {
+			return corecharm.Platform{}, false, err
+		}
+		platform = p
+		platStrings.Add(p.String())
+	}
+
+	return platform, platStrings.Size() == 1, nil
+}
+
+func (v *deployFromRepositoryValidator) resolveCharm(curl *charm.URL, requestedOrigin corecharm.Origin, force, usedModelDefaultBase bool) (*charm.URL, corecharm.Origin, error) {
+	repo, err := v.getCharmRepository(requestedOrigin.Source)
+	if err != nil {
+		return nil, corecharm.Origin{}, errors.Trace(err)
+	}
+
+	resultURL, resolvedOrigin, supportedSeries, resolveErr := repo.ResolveWithPreferredChannel(curl, requestedOrigin)
+	if charm.IsUnsupportedSeriesError(resolveErr) {
+		if !force {
+			msg := fmt.Sprintf("%v. Use --force to deploy the charm anyway.", resolveErr)
+			if usedModelDefaultBase {
+				msg += " Used the default-series."
+			}
+			return nil, corecharm.Origin{}, errors.Errorf(msg)
+		}
+	} else if resolveErr != nil {
+		return nil, corecharm.Origin{}, errors.Trace(resolveErr)
+	}
+
+	// The charmhub API can return "all" for architecture as it's not a real
+	// arch we don't know how to correctly model it. "all " doesn't mean use the
+	// default arch, it means use any arch which isn't quite the same. So if we
+	// do get "all" we should see if there is a clean way to resolve it.
+	if resolvedOrigin.Platform.Architecture == "all" {
```

Comment:
Just wanted to comment that the charmhub sending us "messages" like this in fields (just because the field is a string) is really not a good practice. This should be a flag something like `anyArch = true`.